In [133]:
import glob
import pandas as pd
import numpy as np
from utils.statistics import statistics

In [134]:
!ls data/appalti | head -4

"D.BIRAGO"_ISTITUTO_COMPRENSIVO_2015-01-15.csv
2^_CIRCOLO_DIDATTICO_STATALE_(GARIBALDI)_-_ALTAMURA_(BA)_2015-01-14.csv
81004070314_-_ISTITUTO_COMPRENSIVO_"G.RANDACCIO"_MONFALCONE_2015-01-21.csv
A.S.P._COMUNI_MODENESI_AREA_NORD_2014-12-31.csv


In [135]:
folder_path =r'data/appalti'
allFiles = glob.glob(folder_path + "/*.csv")
frame = pd.DataFrame()
list_ = []
for file_ in allFiles:
    df = pd.read_csv(file_, sep=",")
    list_.append(df)
frame = pd.concat(list_)

In [136]:
frame.head()

cig codicefiscale_prop                    denominazione  \
0  5622811A6A        80004480549  "D.BIRAGO" ISTITUTO COMPRENSIVO   
1  58730989C8        80004480549  "D.BIRAGO" ISTITUTO COMPRENSIVO   
2  Z030DAFD12        80004480549  "D.BIRAGO" ISTITUTO COMPRENSIVO   
3  Z0909E5CE3        80004480549  "D.BIRAGO" ISTITUTO COMPRENSIVO   
4  Z090EC70B9        80004480549  "D.BIRAGO" ISTITUTO COMPRENSIVO   

                                             oggetto  \
0           Saldo fatt. n. V1/0015186 DEL 31-03-2014   
1  Saldo fattura per servizi di ripristino funzio...   
2                Saldo fattura n. 57/14 del 26-05-14   
3  CONTRATTO DI ASSISTENZA ANNUALE HARDWARE 01-09...   
4                   Saldo fatt. n. 212 dell'11-04-14   

                                  scelta_contraente  \
0  23-AFFIDAMENTO IN ECONOMIA - AFFIDAMENTO DIRETTO   
1  23-AFFIDAMENTO IN ECONOMIA - AFFIDAMENTO DIRETTO   
2  23-AFFIDAMENTO IN ECONOMIA - AFFIDAMENTO DIRETTO   
3  23-AFFIDAMENTO IN ECONOMIA - AFFIDAMENTO DIRETTO   
4  23-AFFIDAMENTO IN ECONOMIA - AFFIDAMENTO DIRETTO   

                          partecipanti  aggiudicatari  importo_aggiudicazione  \
0                          02884150588            NaN                48466.62   
1                          02884150588            NaN                 6114.51   
2                          02790150540            NaN                   98.36   
3                          02331650545            NaN                 2344.63   
4  00716420559,02331650545,04552920482            NaN                 1779.16   

  data_inizio   data_fine  importo_somme_liquidate  
0  2014-04-15  2014-12-31                 48466.62  
1  2014-12-31         NaN                     0.00  
2  2014-05-30  2014-06-04                    98.36  
3  2013-09-28  2013-12-31                   857.00  
4  2014-05-06  2014-11-04                    96.58

## Cardinalita e dimensionalità

In [137]:
print ("Il dataset presenta",frame.shape[0],"record e",frame.shape[1],'attributi')

Il dataset presenta 25391 record e 11 attributi


## Controllo della sintassi e attributi

### L'attributo `cig` è sempre nella stessa forma?
L'attributo `cig` (es. X4F0FED52C,58730989C8,Z3612A45) sembra essere caratterizzato da  8-10 caratteri esclusivamente numeri e lettere.

Di seguito si esegue un **controllo di lunghezza** e **un controllo di caratteri ammessi** per evidenziarie eventuali mancanze sintattiche

#### Controllo di lunghezza caratteri

In [150]:
frame['len_cig']=frame['cig'].apply(lambda x: len(str(x)))
frame['len_cig'].value_counts()

10    25240
9       131
1         9
8         5
7         2
6         2
3         1
2         1
Name: len_cig, dtype: int64

In [153]:
# Uncomment for see
frame.query('len_cig < 8')

Int64Index([0, 161, 9, 48, 51, 95, 107, 108, 157, 234, 5, 7, 8, 9, 0], dtype='int64')

In [140]:
frame.query('len_cig < 8')

cig codicefiscale_prop  \
0          0        84503340154   
161   Z2E105          953050762   
9         \n          706280138   
48        \n          706280138   
51        \n          706280138   
95        \n          706280138   
107       \n          706280138   
108       \n          706280138   
157       \n          706280138   
234       \n          706280138   
5        000        80002800458   
7         01        80002800458   
8     000001        80002800458   
9    ZCA0E7B        97028260780   
0    0000000        80008640528   

                                         denominazione  \
0                           COMUNE DI BOFFALORA D'ADDA   
161                              COMUNE DI PIETRAGALLA   
9                                     Comune di Rogeno   
48                                    Comune di Rogeno   
51                                    Comune di Rogeno   
95                                    Comune di Rogeno   
107                                   Comune di Rogeno   
108                                   Comune di Rogeno   
157                                   Comune di Rogeno   
234                                   Comune di Rogeno   
5                         ISTITUTO COMPRENSIVO MASSA 6   
7                         ISTITUTO COMPRENSIVO MASSA 6   
8                         ISTITUTO COMPRENSIVO MASSA 6   
9                      ISTITUTO COMPRENSIVO ROSSANO II   
0    ISTITUTO TECNICO STATALE COMMERCIALE E PER GEO...   

                                               oggetto  \
0    BANDO DI GARA CON PROCEDURA APERTA\nPER L'AFFI...   
161                 FORNITURA  E INSTALLAZIONE  SERVER   
9    IMPEGNO E LIQUIDAZIONE RIMBORSO QUOTA DI TRASP...   
48            LIQUIDAZIONE COMPENSO PER V.A.S. DEL PGT   
51   LIQUIDAZIONE  QUOTE PER SERVIZIO RIFIUTI AL CO...   
95   SERVIZIO DI ASSISTENZA EDUCATIVA SCOLASTICA A....   
107  IMPEGNO DI SPESA E CONTESTUALE LIQUIDAZIONE PE...   
108  IMPEGNO DI SPESA E CONTESTUALE LIQUIDAZIONE PE...   
157  IMPEGNI DI SPESA PER ACQUISTO PRODOTTI PER LE ...   
234  APPROVAZIONE DEL PROTOCOLLO D'INTESA PER LA RE...   
5                               COLLABORAZIONE PLURIMA   
7       CONTRATTO DI PRESTAZIONE D'OPERA PROFESSIONALE   
8     CONTRATTO PER PRESTAZIONE AUTONOMA PROFESSIONALE   
9                           INCARICO RESPONSABILE RSPP   
0    saldo fattura n.  4947/S del 01/12/2014 - acqu...   

                                     scelta_contraente  \
0                                  01-PROCEDURA APERTA   
161                07-SISTEMA DINAMICO DI ACQUISIZIONE   
9     23-AFFIDAMENTO IN ECONOMIA - AFFIDAMENTO DIRETTO   
48    23-AFFIDAMENTO IN ECONOMIA - AFFIDAMENTO DIRETTO   
51   25-AFFIDAMENTO DIRETTO A SOCIETA' RAGGRUPPATE/...   
95    23-AFFIDAMENTO IN ECONOMIA - AFFIDAMENTO DIRETTO   
107   23-AFFIDAMENTO IN ECONOMIA - AFFIDAMENTO DIRETTO   
108   23-AFFIDAMENTO IN ECONOMIA - AFFIDAMENTO DIRETTO   
157   23-AFFIDAMENTO IN ECONOMIA - AFFIDAMENTO DIRETTO   
234   23-AFFIDAMENTO IN ECONOMIA - AFFIDAMENTO DIRETTO   
5     23-AFFIDAMENTO IN ECONOMIA - AFFIDAMENTO DIRETTO   
7                                  01-PROCEDURA APERTA   
8                                  01-PROCEDURA APERTA   
9    03-PROCEDURA NEGOZIATA PREVIA PUBBLICAZIONE DE...   
0    04-PROCEDURA NEGOZIATA SENZA PREVIA PUBBLICAZI...   

                partecipanti  aggiudicatari  importo_aggiudicazione  \
0                        NaN            NaN                    0.00   
161              01101030763            NaN                 2800.00   
9           CLMMHL72D62B081R            NaN                  100.00   
48               91002210135            NaN                 1500.00   
51               91002210135            NaN                29508.36   
95               00623530136            NaN                14300.00   
107              91012870159            NaN                  201.30   
108              91012870159            NaN                  100.00   
157              04754

#### Controllo di caratteri ammessi
All'interno dell'attributo `cig` sono ammessi sono lettere e numeri? O esistono altri caratteri stani?

In [141]:
len_cig[len_cig['cig'].str.match('^[a-zA-Z0-9]*$')==False]

cig  len_cig
108   Z42 FA77DB       10
114   Z81 FD9354       10
123   Z691 2F352       10
40    CIG ZBA0E2       10
45    NON NECESS       10
62    NON NECESS       10
63    NON NECESS       10
102   NON NECESS       10
123   NON NECESS       10
125   NON NECCES       10
130   NON NECESS       10
131   NON NECESS       10
132   NON NECESS       10
137   NON NECESS       10
168   NON NECESS       10
174   NON NECESS       10
175   NON NECESS       10
186   NON NECESS       10
201   NON NECESS       10
94    VD. OGGETO       10
8     CIG ZCD09D       10
9     CIG ZCD09D       10
68   \tZB20B94CF       10
145  \tZ7809604F       10
109   ZE6120/C12       10
71     Z561241A0       10
72     Z3E124E56       10
499    Z1E0C29FB       10
0      482107904       10
94    ZBF B8B590       10
0      ZE10AD6C7       10
38     Z161016D4       10
157    ZD511CA8E       10

In [142]:
#frame[len_cig['cig'].str.match('^[a-zA-Z0-9]*$')==False].head()

Sono presenti `cig` con spazi,`\t` e `\n`, non di meno sono presenti elementi con dicitura `NON NECESS`, si procede al trattamento con la rimozione dei caratteri, suddetti, non conformi e al settaggio a Missing Values degli elementi valorizzati a `NON NECESS`.

In [149]:
frame['cig']=frame['cig'].apply(lambda x: x.replace('\n', '',).replace('\t','').replace(' ',''));frma


TypeError: Cannot change data-type for object array.

In [143]:
#def conforma_sintassi_cig(x):
#    return np.char.replace(np.char.replace(x, '\t', ''), ' ','')

#conforma_sintassi_cig(frame['cig'])    
frame['url']=df_suc['url'].apply(lambda x: x.strip('\n').strip(' ').replace(" ", "").replace("\t", ""))

TypeError: string operation on non-string array

### Attributo `cig`

L'attributo non risulta essere univoco, molti valori sono caratterizzati a `0000000000` che sembrerebbe riconducibile a missing value

In [37]:
frame.cig.value_counts(dropna=False).head(5)

0000000000    770
X4F0FED52C    111
XE50FEEFE0     74
ZA2118A16E     58
5609937A74     23
Name: cig, dtype: int64

In [38]:
frame.cig.value_counts(dropna=False).tail(5)

Z090BC7082    1
Z7E1046F85    1
6038406A42    1
5686181101    1
0             1
Name: cig, dtype: int64

#### Controllo record duplicati
Come si evince dalla seguente query, i record risultano duplicati. Tuttavia la duplicazione non si riferisce a tutto il record ma differisce soltanto negli attributi: `oggetto`, `partecipanti`, e infine nell'`importo_aggiudicazione`.

A titolo esemplificativo, basta osservare il primo e il secondo record, mentre il primo riporto in oggetto "CIG X4F0FED52C - LIQUIDAZIONE I^ PARTE FATTURA", Il secondo ricorso in oggetto "CIG X4F0FED52C - LIQUIDAZIONE FATTURA" mentre il primo ha un importo in aggiudicazione di €144.20 il secondo €34.65.

In [49]:
frame.query('cig ==\'X4F0FED52C\'').head(2)

cig codicefiscale_prop      denominazione  \
41  X4F0FED52C         1375760012  Comune di Bruzolo   
42  X4F0FED52C         1375760012  Comune di Bruzolo   

                                           oggetto  \
41  CIG X4F0FED52C - LIQUIDAZIONE I^ PARTE FATTURA   
42           CIG X4F0FED52C - LIQUIDAZIONE FATTURA   

                                   scelta_contraente             partecipanti  \
41  23-AFFIDAMENTO IN ECONOMIA - AFFIDAMENTO DIRETTO  02356770988,02356770988   
42  23-AFFIDAMENTO IN ECONOMIA - AFFIDAMENTO DIRETTO  02356770988,02356770988   

    aggiudicatari  importo_aggiudicazione data_inizio data_fine  \
41            NaN                  144.20         NaN       NaN   
42            NaN                   34.65         NaN       NaN   

    importo_somme_liquidate  
41                      0.0  
42                      0.0

Infine mentre i precedenti record mostrano due partecipanti "02356770988,02356770988" i seguenti record pur riferendosi allo stesso `cig`, hanno un solo partecipante "02356770988".

In [50]:
frame.query('cig ==\'X4F0FED52C\'').tail(2)

cig codicefiscale_prop      denominazione  \
633  X4F0FED52C         1375760012  Comune di Bruzolo   
634  X4F0FED52C         1375760012  Comune di Bruzolo   

                                             oggetto  \
633  CIG X4F0FED52C - LIQUIDAZIONE II^ PARTE FATTURA   
634  CIG X4F0FED52C - LIQUIDAZIONE II^ PARTE FATTURA   

                                    scelta_contraente partecipanti  \
633  23-AFFIDAMENTO IN ECONOMIA - AFFIDAMENTO DIRETTO  02356770988   
634  23-AFFIDAMENTO IN ECONOMIA - AFFIDAMENTO DIRETTO  02356770988   

     aggiudicatari  importo_aggiudicazione data_inizio data_fine  \
633            NaN                  144.20         NaN       NaN   
634            NaN                  286.76         NaN       NaN   

     importo_somme_liquidate  
633                      0.0  
634                      0.0

**dal controllo dei duplicati è emerso** che ogni record rappresenta una movimentazione di denaro

#